example generation for typespec

tadelesh opened the pull request:
resolve: https://github.com/Azure/autorest.go/issues/1358

basic design concept:
1. add example in go code model
2. map tcgc example model into go code model, do the mapping in the same code generator adapter
3. use go code model to generate example files, do the generation in the same code generator processor
4. TODO: need to migrate swagger go test generator logic, adapt swagger test modeler example model to go code model, but currently low priority.

jhendrixMSFT commented on packages/codemodel.go/src/client.ts:
```diff
@@ -181,6 +184,7 @@ export class Method implements Method {
     this.name = name;
     this.naming = naming;
     this.parameters = new Array<param.Parameter>();
+    this.examples = [];
```
`this.examples = new Array<MethodExample>();`

jhendrixMSFT replied:
See linter errors for more changes.

jhendrixMSFT commented on packages/typespec-go/src/tcgcadapter/adapter.ts:
```diff
@@ -15,14 +15,14 @@ const headerText = `Copyright (c) Microsoft Corporation. All rights reserved.
   }
 
-  const sdkContext = tcgc.createSdkContext(context);
+  const sdkContext = await tcgc.createSdkContext(context);
```
Why was this function made async?

tadelesh replied:
this is from the latest breaking change of tcgc 0.44.2

tadelesh replied:
move to this pr: https://github.com/Azure/autorest.go/pull/1411

jhendrixMSFT commented on packages/typespec-go/src/tcgcadapter/types.ts:
```diff
@@ -3,13 +3,13 @@
 
-import * as naming from '../../../naming.go/src/naming.js';
-import * as go from '../../../codemodel.go/src/index.js';
+import { uncapitalize } from '@azure-tools/codegen';
```
In the future, it would be ideal to omit such refactoring from PRs with functional changes to reduce the delta, making it easier for reviewers to focus on the actual changes.

tadelesh replied:
sure. do you think i need to revert it from this pr?

jhendrixMSFT replied:
Agreed on the consistency. I will work on updating CI to enforce this.

tadelesh replied:
reverted all format change.

jhendrixMSFT commented on packages/typespec-go/src/tcgcadapter/types.ts:
```diff
@@ -149,20 +149,20 @@ export class typeAdapter {
   // the operation is idempotent, so getting the same type multiple times
   // returns the same instance of the converted type.
-  getPossibleType(type: tcgc.SdkType, elementTypeByValue: boolean, substituteDiscriminator: boolean): go.PossibleType {
+  getPossibleType(type: tcgc.SdkType, elementTypeByValue: boolean, substituteDiscriminator: boolean, isMulti = false): go.PossibleType {
```
What does `isMulti` mean/do?

tadelesh replied:
it come from the new multi-part feature of typespec and related change from tcgc, also see https://github.com/Azure/typespec-azure/pull/1237. i think i should move the upgrade of new tcgc to a new pr which could make this pr focus on example only.

tadelesh replied:
will remove after https://github.com/Azure/typespec-azure/pull/1251 release

jhendrixMSFT commented on packages/typespec-go/src/tcgcadapter/types.ts:
```diff
@@ -352,16 +353,6 @@ export class typeAdapter {
         this.types.set(decimalKey, decimalType);
         return decimalType;
       }
-      case 'eTag': {
```
Was `eTag` and the other removed types converted to `string`?

tadelesh replied:
same from: https://github.com/Azure/typespec-azure/pull/1237

tadelesh replied:
move to: https://github.com/Azure/autorest.go/pull/1411

jhendrixMSFT replied:
The motivation for the change makes sense. However, my concern here is that we no longer have a predefined union of type names, so we don't know the possible "sub-type" when writing the code and are left with magic strings.

If it's not possible to define a union for these sub-types, can we at least define some constants, or document the possible sub-types someplace? Maybe it's already there in some form and I just missed it?

tadelesh replied:
sure. created an issue: https://github.com/Azure/typespec-azure/issues/1260.

ArcturusZhang replied:
So the ask here is to document the `crossLanguageDefinitionId`s of those azure core types?
I do not think we should include a union in TCGC to enable that - the initial motivation of moving them out from the built-in type is trying not to make TCGC strongly bond with azure when we do not have to.
I think we could put those crossLanguageDefinitionId in a document for azure types.

tadelesh replied:
correct. we could put the doc here: https://azure.github.io/typespec-azure/docs/howtos/DataPlane%20Generation%20-%20DPG/07tcgcTypes

jhendrixMSFT replied:
Can you please add a link to this from the tcgc code so it's easy to find? Who will own maintaining it over time?

tadelesh replied:
after the init version, i could help to keep it up-to-date. the scalar types that need special handling should be limited.

jhendrixMSFT commented on packages/typespec-go/src/tcgcadapter/clients.ts:
```diff
@@ -263,13 +265,15 @@ export class clientAdapter {
     }
 
-    this.adaptMethodParameters(sdkMethod, method);
+    this.adaptMethodParameters(sdkMethod, method, paramMapping);
```
It looks like `adaptMethodParameters` is populating `paramMapping` yes? If so, can we change this so `adaptMethodParameters` returns it instead of using an out param?

tadelesh replied:
changed.

jhendrixMSFT commented on packages/codegen.go/src/clientFactory.ts:
```diff
@@ -87,3 +78,17 @@ export async function generateClientFactory(codeModel: go.CodeModel): Promise<st
+  }
+  allClientParams.sort(sortParametersByRequired);
+  return allClientParams;
+}
```
Missing new-line.

tadelesh replied:
added

jhendrixMSFT commented on packages/codegen.go/src/clientFactory.ts:
```diff
@@ -87,3 +78,17 @@ export async function generateClientFactory(codeModel: go.CodeModel): Promise<st
   return result;
 }
+
+export function getAllClientParameters(codeModel: go.CodeModel): Array<go.Parameter> {
```
This should probably move to `helpers.js` if it needs to be used in multiple places.

tadelesh replied:
moved.

jhendrixMSFT commented on packages/codegen.go/src/example.ts:
```diff
@@ -0,0 +1,404 @@
+}
+
+// Creates the content for all <example>.go files
+export async function generateExamples(codeModel: go.CodeModel): Promise<Array<ExampleContent>> {
```
Do you think there's value in adding an opt-out flag to disable example generation? I'm not advocating for or against, just wondering.

tadelesh replied:
added.

tadelesh commented:
@jhendrixMSFT do you have further concern for this pr?

jhendrixMSFT commented:
I had a few questions that were never answered (see above).

tadelesh commented:
> I had a few questions that were never answered (see above).

oh, i just changed the code according to the comment and resolved the comment directly. added comment.
